File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
class Solution {
public:
/**
   [3, 2, 4, 6] ==> [1,3,6,24]
                    [48,24,6,1]
                 ==> []

    space is O(n) and runtime is O(n)
*/

    vector<int> productExceptSelf(vector<int>& nums) {
        int changed_value = 1;
        vector<int> output_vals(nums.size(),1);
        for (int i = 0; i < nums.size(); i++){
            output_vals[i] = changed_value;
            changed_value*=nums[i];
        }
        changed_value = 1;
        for (int j = nums.size() - 1; j >= 0; j--){
            output_vals[j]*=changed_value;
            changed_value *= nums[j];
        }
        return output_vals;

    }
};

File: Data Structures & Algorithms/daily-temperatures/submission-2.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        # Approach #1: runtime is O(n^2) and space is O(n)
        result = [0]*len(temperatures)
        stack = []
        for r in range(len(temperatures)):
            while stack and stack[-1][0] < temperatures[r]:
                result[stack[-1][1]] = r - stack[-1][1]
                stack.pop()
            stack.append([temperatures[r],r]) 
        return result
        